### PR TITLE
[Input] Windows input fixes

### PR DESCRIPTION
**Pull request by Eideren**
Fixing issue #159 
While this first push addresses the main issue mentioned there it makes the scene view hard to use because of the missing events mentioned lower in that same issue.
I'm investigating it, if you have any clue feel free to share !

n.b.:Those `Console.WriteLine` will be gone once I'm done of course.

### CODE REVIEW

File: sources/engine/Xenko.Input/InputManager.State.cs [hunk 1]
```diff
@@ -110,7 +110,7 @@ public IReadOnlySet<Keys> PressedKeys
         {
             get
             {
-                if (!HasMouse) return NoKeys;
+                if (!HasKeyboard) return NoKeys;
```
Comment (Kryptos-FR) [inline sources/engine/Xenko.Input/InputManager.State.cs]: :open_mouth: good catch!

File: sources/engine/Xenko.Input/Windows/InputSourceWinforms.cs [hunk 2]
```diff
@@ -165,6 +171,20 @@ private static WinFormsKeys GetCorrectExtendedKey(WinFormsKeys virtualKey, long
...
+        /// <br/>This function is used to find out if the received message is a repeated KEYDOWN.
+        /// </summary>
+        /// <param name="lParam">lParam of the KEYDOWN message</param>
+        /// <returns><c>True</c> if this message is a repeated KeyDown, <c>false</c> if it's an actual keydown</returns>
+        private static bool MessageIsDownAutoRepeat(IntPtr lParam)
+        {
+            // According to the microsoft docs on WM_KEYDOWN
+            // (https://docs.microsoft.com/en-us/windows/desktop/inputdev/wm-keydown)
+            // The second to last bit is 0 when the last keyboard message was up and 1 if it was already down
+            return ((int)lParam & (1 << 30)) != 0;
```
Comment (Kryptos-FR) [inline sources/engine/Xenko.Input/Windows/InputSourceWinforms.cs]: @xen2 should that be enclosed by `unchecked`? `IntPtr` can be 32-bis or 64-bits depending on the platform if I'm not mistaken.

Comment (Eideren) [reply sources/engine/Xenko.Input/Windows/InputSourceWinforms.cs]: You're right, I'll just use a long since it's already used in the other function.

**Comment by Eideren** [comment]
So, the input system is fine, native WndProc just fails to send us all of the user's inputs. 
It's most likely related to the UI element bound to that function since that's really the only other variable in this equation and it also fails to get some inputs.

We could use direct or raw input and call it a day but I'm pretty sure that this issue is larger than just input related and could solve a bunch of those on #94 .

What's going on with that UI element ?

**Comment by Kryptos-FR** [comment]
That's out of my expertise :sweat_smile: let's ask @xen2

**Comment by xen2** [comment]
@Eideren Interesting find! Unfortunately I have no idea why it fixes the issue. Pinging @xoofx :)

**Comment by xoofx** [comment]
Can't tell why (don't remember the codebase of the winform in Xenko), but before hacking with RawInput here, I would verify the messages at SPY++ level on the form to make sure that it is not WinForms early filtering something badly. In that case, it might be possible to find a solution within it.

**Comment by Eideren** [comment]
Spy++ does report that we receive all of our keyup/down messages.

Now just to triple verify our filtering, I took a look at how the forms are used and found out that we don't actually use the common windows forms way of initializing interfaces, we pull and push messages to the UI through [this function](https://github.com/xenko3d/xenko/blob/1cc4d2761db038ebe89f04c1787b572b9ebc6dc0/sources/engine/Xenko.Games/Desktop/WindowsMessageLoop.cs#L126) instead ; so I removed all of the overridden key functions on the form and turned `UseApplicationDoEvents` to true to use the standard `Application.DoEvents()` in place of our custom version and ... nothing changed.

'Implemented a [PreProcessMessage](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.control.preprocessmessage?view=netframework-4.7.2) method on the form and nope, same results there.

I then tried to launch our scene view through Application.Run or form.Show, I'm really grasping at straws here, but both of them couldn't act properly.

There doesn't seem to be any kind of concrete source/explanation as to what RegisterRawInputDevices does either, besides the fact that it [enables raw input reception for applications](https://docs.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-registerrawinputdevices).

I can't think of anything else, if you have any idea at this point, I'm all ears.

**Comment by Eideren** [comment]
Alright, since this is getting nowhere and we still have to at least provide a work around for this issue I pushed it anyway.
Would you like me to create a new issue to replace the work around or does the 'TODO' inside the summary for the function suffice ?

**Comment by xen2** [comment]
Yes I am good with using `RegisterRawInputDevices` for now considering how much improved the situation would be.
TODO with a quick explanation and link to this conversation should be enough for now.
Thanks!

**Comment by xen2** [comment]
Thanks!

**Pull request merged**